collectively-exhaustive-condition-part

danielczinege opened the pull request:
This checker should also be done. It does not detect the example from our defect table: `(day % a == 0 and day % b != 0) or (day % a == 0 and day % b == 0)`, because the `guess_type()` function cannot yet tell the type of `day % a`.

Also it only found two examples in the whole umime_classic and when I turned off the type checking with `guess_type()` ( commented it out from initialize_variables() function), then it found additional 2 examples, where both of them should start being detected when `guess_type()` starts guessing type of `m % n`.

I tried to make it such that if there is something like: `(A and B) or (C and A) or (B and D)`, where `B or C != True`, but `A or D ≡ True`, then it would suggest that `(A and B) or (B and D)` can be replaced with `B`. (ie. even when they are not necessarily next to each other - of course I first check for purity).

I have also added and used function `are_same_expression(expr1, expr2)` into `utils.py` with a simple implementation, by just comparing their `.as_string()`. I guess that you would use this function for the `duplicate condition part` checker when you will be implementing it, but you might enhance it by also dealing with commutativity, ...

danielczinege committed:
checkers(collectively-exhaustive-condition-part): first version

danielczinege committed:
refactor: _find_collectively_exhaustive_condition_part()

danielczinege committed:
refactor: further improvements to function _find_collectively_exhaustive_condition_part()

danielczinege committed:
tests: collectively-exhaustive-condition-part

GiraffeReversed committed:
analyses: merge are_identical and are_same_expression

GiraffeReversed commented:
Hm, it seemed to me that the issue is a bit more prevalent when I was going through the data. I am going to look for other defects manually anyway, so we'll see if I find any other cases.
When running the detector on my end, I did not find the two cases in Umíme. Could you add them to tests as files?

I did not improve the type inference in the past weeks as I hoped. I decided to do much larger refactoring. Let us leave the detector as it is now, using `guess_type`, and I will do the improvements when I get to them.

I merged your `are_same_expression` with `are_identical` function defined just above, which did the very same thing. I kept your name, I think it is a bit better :)

danielczinege committed:
checkers: change code to R6225 and make new tests

danielczinege commented:
Yes, in the umime_classic that I got from you it found only one case, regardless if I let the guess_type() be there or not. But in the data from IV127 course it found the two cases I just added to the test submodule when I used guess_type(). And when I turned it off it found these:
(the ones with guess_type())
'`i % 2 == 0 and j % 2 == 1 or i % 2 == 1 and j % 2 == 1`' can be replaced with '`j % 2 == 1`'
'`l % 2 == vyska % 2 and k % 2 == 0 or l % 2 == vyska % 2 and k % 2 == 1`' can be replaced with '`l % 2 == vyska % 2`'

(the ones without guess_type())
'`heads % 2 == 1 and pigs % 2 == 0 or heads % 2 == 0 and pigs % 2 == 0`' can be replaced with '`pigs % 2 == 0`'
'`p % k == 0 and p > k * 4 or p % k != 0 and p > k * 4`' can be replaced with '`p > k * 4`'
'`(rows == 0 or rows % 2 == 0) and (cols + rows) % 2 == 0 or rows % 2 != 0 and (cols + rows) % 2 == 0`' can be replaced with '`(cols + rows) % 2 == 0`'

So actually it was additional 3 cases. I am not sure why I thought it was only two, but also not all of them will be fixed even when guess_type() will guess the type of `m%n`, also don't know where I got that before. :)

I hope I didn't mess something up with the submodule. When I added the two files it said I have detached HEAD, so I checked out to main and cherry picked the commit I made with the two files and pushed it there.

GiraffeReversed commented:
The test files came through just fine, thanks. I was confused as my analysis did not find any files, but now I see what happened. I used a slightly different dataset, and the two files contained a mix of tabs and spaces, which counts as syntax error, so they weren't parsed and analyzed.

danielczinege committed:
checkers: remove warning about checkers using Z3 possibly giving incorrect suggestions when % or // is used with floats, because we use guess_type()

danielczinege commented:
I realised I have to remove the warning on this branch as well.

danielczinege committed:
fix: remove 'print' from PURE_FUNCTIONS
